server/src/main/java/org/elasticsearch/cluster/routing/IndexShardRoutingTable.java

```diff
@@ -62,7 +62,7 @@
     IndexShardRoutingTable(ShardId shardId, List<ShardRouting> shards) {
         this.shardId = shardId;
         this.shuffler = new RotationShardShuffler(Randomness.get().nextInt());
-        this.shards = Collections.unmodifiableList(shards);
+        this.shards = List.copyOf(shards);
```

Comment:
I wonder if we should replace all `Collections.unmodifiableList` calls with `List.copyOf` across the codebase to avoid copying un-modifiable lists? Same with sets and maps.

It would be great to introduce some check to prevent new usages of `Collections.empty/unmodifiable` that could be done with `List.of/Set.of`Map.of`

Reply:
I think it's hard to automatically identify all cases like this one. There's lots of legitimate uses of the unmodifiable wrappers too when copying would be needlessly expensive because e.g. lists are long.
Also, in most cases this kind of thing is probably irrelevant. This one is a very particular case where nested iteration + a large number of short lists came together. In most cases you'll probably not notice this kind of wrapping because it gets inlined and/or eliminated by escape analysis I'd assume.

Reply:
Looks like `IndexShardRoutingTable` is only created from `build()`. I would suggest to remove the `copyOf` from the build, this way you could remove one more instanceof call then. Since the `copyOf` will still be in constructor this should remain safe

Reply:
Btw, in the recent versions of the JDK calling `unmodifiableList` on an `UnmodifiableList` is a no-op: https://github.com/openjdk/jdk/commit/dbef0ec95d0fd686b63fa4a873cc3f580719ba30

Reply:
Correct, but unfortunately this check is not working for collection produced by `List.of` or `List.copyOf` even though those are immutable as well. 

So in short:
* `List.copyOf(List.copyOf(something))` does single copy
* `Collections.unmodifiableList(Collections.unmodifiableList(something))` also wraps once
but
* `List.copyOf(Collections.unmodifiableList(something))` and `Collections.unmodifiableList(List.copyOf(something))` do not have any optimizations.